### PR TITLE
Manage ec2

### PR DESCRIPTION
**Pull request by krjackso**

### CODE REVIEW

File: lib/aws_extensions/ec2/Instance.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    module Instance
+
+      # Public: Returns the value of the "Name" tag for the Instance
+      def name
+        self.tags.select { |tag| tag.key == "Name" }.first.value
+      rescue
+      	nil
+      end
+
+      # Public: Returns an rray of the block device mappings that are not for the root device
```
Comment (msiebert) [inline lib/aws_extensions/ec2/Instance.rb]: missing an a here. Either put the a in or use an apostrophe, as in 'rray, the cowboy coder's version of array.

File: lib/common/BaseLoader.rb [hunk 2]
```diff
@@ -10,12 +10,13 @@ module BaseLoader
       #
       # dir               - the directory to load resources from
       # individual_loader - the function that loads a resource from each file name
+      # json              - indicates if the resources are in json format
```
Comment (msiebert) [inline lib/common/BaseLoader.rb:12]: you should put this parameter above `individual_loader`

File: lib/ec2/managers/EbsManager.rb [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+        # Migrate any volumes that have already been grouped
+        vol_groups = Hash[EC2.group_ebs_volumes.map { |group_name, cumulus_group| [group_name, EbsMigrationData.new(cumulus_group, nil)] }]
+
+        # Use the instance name to group volumes if they do not have a group.  Anything
+        # not attached should not get migrated
+        migratable_vols = EC2.ebs_volumes.select { |vol| vol.group.nil? and !vol.attachments.empty? and vol.attached? }
+        instance_grouped = migratable_vols.group_by do |vol|
+          attachments = vol.attachments.select { |att| att.state == "attached" || att.state == "attaching" }
+          attachments.first.instance_id
```
Comment (msiebert) [inline lib/ec2/managers/EbsManager.rb:74]: Do you need a rescue here for the case that nothing is attached?

File: lib/ec2/managers/InstanceManager.rb [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+      end
+
+      def local_resources
+        @local_resources ||= Hash[InstanceLoader.instances.map { |local| [local.name, local] }]
+      end
+
+      def aws_resources
+        @aws_resources ||=
+          EC2::named_instances
+            .reject { |name, i| AutoScaling::instance_ids.include?(i.instance_id) }
```
Comment (msiebert) [inline lib/ec2/managers/InstanceManager.rb:38]: Make sure to document this behavior. I assume it's obvious to most people that this is desired, but you might help out that one dumb dude who doesn't think about it.

File: lib/ec2/managers/InstanceManager.rb [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+          cumulus_instance = InstanceConfig.new(name).populate!(instance, user_data_file)
+
+          json = JSON.pretty_generate(cumulus_instance.to_hash)
+          File.open("#{instances_dir}/#{name}.json", "w") { |f| f.write(json) }
+        end
+
+      end
+
+      def create(local)
+        #######################################
```
Comment (msiebert) [inline lib/ec2/managers/InstanceManager.rb]: Is this a todo?

File: lib/ec2/managers/InstanceManager.rb [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+            }
+          end,
+          iam_instance_profile: {
+            arn: profile_arn
+          },
+          ebs_optimized: local.ebs_optimized
+        }).instances.first
+
+        # Wait until the instance is running then attach volumes
+        print "Waiting for instance to run"
```
Comment (msiebert) [inline lib/ec2/managers/InstanceManager.rb]: why print and not puts?

File: lib/ec2/managers/EbsManager.rb (hunk 3 above)
Comment (krjackso) [inline lib/ec2/managers/EbsManager.rb:74]: the line above with `migratable_vols` ensures there is at least one attachment with the correct state.  It's a little obscure that the check is being done I know, and this is why doing things like select+map at the same time would be so nice

File: lib/ec2/managers/InstanceManager.rb (hunk 4 above)
Comment (krjackso) [inline lib/ec2/managers/InstanceManager.rb:38]: Yeah it's in the documentation that you will see shortly

File: lib/aws_extensions/ec2/Instance.rb (hunk 1 above)
Comment (krjackso) [inline lib/aws_extensions/ec2/Instance.rb]: it's the opposite of yarr

File: lib/common/BaseLoader.rb (hunk 2 above)
Comment (krjackso) [inline lib/common/BaseLoader.rb:12]: ah good find. Turns out you can't have parameters after a block parameter or something...

**Pull request merged**